File: solutions/155.min-stack.java
/*
 * @lc app=leetcode id=155 lang=java
 *
 * [155] Min Stack
 */

// @lc code=start

import java.util.Stack;

class MinStack {
    private Stack<Integer> store = new Stack<>();
    private Stack<Integer> minStore = new Stack<>();

    public MinStack() {
        
    }
    
    public void push(int val) {
        store.push(val);
        int min = minStore.isEmpty() ? val : Math.min(minStore.peek(), val);
        minStore.push(min);
    }
    
    public void pop() {
        store.pop();
        minStore.pop();
    }
    
    public int top() {
        return store.peek();
    }
    
    public int getMin() {
        return minStore.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
// @lc code=end

File: solutions/155.min-stack.py
#
# @lc app=leetcode id=155 lang=python3
#
# [155] Min Stack
#

# @lc code=start
class MinStack:

    def __init__(self):
        self.store = []
        self.minStore = []

    def push(self, val: int) -> None:
        self.store.append(val)
        if not self.minStore or val < self.minStore[-1]:
            self.minStore.append(val)
        else:
            self.minStore.append(self.minStore[-1])

    def pop(self) -> None:
        self.store.pop()
        self.minStore.pop()

    def top(self) -> int:
        return self.store[-1]

    def getMin(self) -> int:
        return self.minStore[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end

